# Contents

- [Built-in data structures](#Built-in-data-structures)
- [Built-in functions](#Built-in-functions)
- [Miscellaneous important concepts](#Miscellaneous-important-concepts)
    - [Special function parameters](#Special-function-parameters)
    - [Variable scope, global and local](#Variable-scope,-global-and-local)
    - [With Statement Context Managers](#With-Statement-Context-Managers)
    - [Errors and Exception Handling](#Errors-and-Exception-Handling)

# Built-in data structures

## Numbers

In [4]:
# Division
7/4

1.75

In [2]:
hex(512)

'0x200'

In [5]:
bin(512)

'0b1000000000'

In [6]:
pow(3,4)

81

In [1]:
pow(3,4,5)

1

In [10]:
round(3,2)

3

In [11]:
round(395,-2)

400

In [12]:
round(3.1415926535,2)

3.14

In [2]:
# Floor Division
7//4

1

## Strings
### Basics

In [ ]:
s="Hello World"
s[::-1]

In [ ]:
s[2::-1]

In [ ]:
letter = 'z'
letter*10

In [5]:
s = 'Hello World concatenate me!'

In [6]:
s.upper()

'HELLO WORLD CONCATENATE ME!'

In [7]:
s.lower()

'hello world concatenate me!'

In [8]:
s.split()

['Hello', 'World', 'concatenate', 'me!']

In [9]:
s.split('W')

['Hello ', 'orld concatenate me!']

In [1]:
s = 'hello world'

In [2]:
s.capitalize()

'Hello world'

In [3]:
s.upper()

'HELLO WORLD'

In [4]:
s.lower()

'hello world'

**Remember, strings are immutable. None of the above methods change the string in place, they only return modified copies of the original string.**

To change a string requires reassignment. 

In [6]:
s = s.upper()
s

'HELLO WORLD'

In [7]:
s = s.lower()
s

'hello world'

In [10]:
s.find('o') # returns the starting index position of the first occurence

4

In [11]:
s.center(20,'z')

'zzzzhello worldzzzzz'

In [12]:
'hello\thi'.expandtabs()

'hello   hi'

In [13]:
s = 'hello'

In [14]:
s.isalnum() #return True if all characters in **s** are alphanumeric

True

In [15]:
s.isalpha() # return True if all characters in s are alphabetic

True

In [16]:
s.islower()

True

In [17]:
s.isspace() #return True if all characters in s are whitespace.

False

In [19]:
s.isupper()

False

In [20]:
s.endswith('o') # essentially the same as a boolean check on s[-1]

True

In [21]:
s.split('e')

['h', 'llo']

In [22]:
s.partition('l') #return a tuple

('he', 'l', 'lo')

### String Formating

There are three ways to perform string formatting.
* The oldest method involves placeholders using the modulo `%` character.
* An improved technique uses the `.format()` string method.
* The newest method, introduced with Python 3.6, uses formatted string literals, called *f-strings*.

#### Formatting with placeholders

In [4]:
print("I'm going to inject %s here." %'something')

I'm going to inject something here.


You can pass multiple items by placing them inside a tuple after the `%` operator.

In [5]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


You can also pass variable names:

In [6]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


Format conversion methods.
It should be noted that two methods <code>%s</code> and <code>%r</code> convert any python object to a string using two separate methods: `str()` and `repr()`. We will learn more about these functions later on in the course, but you should note that `%r` and `repr()` deliver the *string representation* of the object, including quotation marks and any escape characters.

In [7]:
print('He said his name was %s.' %'Fred')
print('He said his name was %r.' %'Fred')

He said his name was Fred.
He said his name was 'Fred'.


As another example, `\t` inserts a tab into a string.

In [8]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')

I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


The `%s` operator converts whatever it sees into a string, including integers and floats. The `%d` operator converts numbers to integers first, without rounding. Note the difference below:

In [9]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


Padding and Precision of Floating Point Numbers
Floating point numbers use the format <code>%5.2f</code>. Here, <code>5</code> would be the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. Let's see some examples:

In [10]:
print('Floating point numbers: %5.2f' %(13.144))

Floating point numbers: 13.14


In [11]:
print('Floating point numbers: %1.0f' %(13.144))

Floating point numbers: 13


In [12]:
print('Floating point numbers: %1.5f' %(13.144))

Floating point numbers: 13.14400


In [13]:
print('Floating point numbers: %10.2f' %(13.144))

Floating point numbers:      13.14


In [14]:
print('Floating point numbers: %25.2f' %(13.144))

Floating point numbers:                     13.14


For more information on string formatting with placeholders visit https://docs.python.org/3/library/stdtypes.html#old-string-formatting

Multiple Formatting
Nothing prohibits using more than one conversion tool in the same print statement:

In [15]:
print('First: %s, Second: %5.2f, Third: %r' %('hi!',3.1415,'bye!'))

First: hi!, Second:  3.14, Third: 'bye!'


#### Formatting with the `.format()` method
A better way to format objects into your strings for print statements is with the string `.format()` method. The syntax is:

    'String here {} then also {}'.format('something1','something2')
    
For example:

In [16]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


The .format() method has several advantages over the %s placeholder method:

 1. Inserted objects can be called by index position:

In [17]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


2. Inserted objects can be assigned keywords:

In [18]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


3. Inserted objects can be reused, avoiding duplication:

In [19]:
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


Alignment, padding and precision with `.format()`
Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

In [20]:
print('{0:8} | {1:9}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))

Fruit    | Quantity 
Apples   |       3.0
Oranges  |        10


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [21]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


You can precede the aligment operator with a padding character

In [22]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [23]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

For more information on the string `.format()` method visit https://docs.python.org/3/library/string.html#formatstrings

#### Formatted String Literals (f-strings)

Introduced in Python 3.6, f-strings offer several benefits over the older `.format()` string method described above. For one, you can bring outside variables immediately into to the string rather than pass them as arguments through `.format(var)`. **No longer necessary to use the key word format**.

In [24]:
name = 'Fred'

print(f"He said his name is {name}.")

He said his name is Fred.


Pass `!r` to get the string representation:

In [25]:
print(f"He said his name is {name!r}")

He said his name is 'Fred'


Float formatting follows `"result: {value:{width}.{precision}}"`

Where with the `.format()` method you might see `{value:10.4f}`, with f-strings this can become `{value:{10}.{6}}`


In [23]:
num = 23.45678
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568


Note that with f-strings, *precision* refers to the total number of digits, not just those following the decimal. **This fits more closely with scientific notation and statistical analysis.** Unfortunately, f-strings do not pad to the right of the decimal, even if precision allows it:

In [24]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


**If this becomes important, you can always use `.format()` method syntax inside an f-string:**

In [25]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


## List 
### Basics

In [1]:
list1 = [1,2,3,4]
list1.pop()
# default pop up the last index. However, in python we can pop any element, e.g. list1.pop(2)


new_list = [1, 2, 3, 4, 5]

#If cannot understand clearly, refer to the Questions.jypnb.
print(new_list.reverse()) #It reverses in place but just return none. So print None
print(new_list) #Because it already reversed in place, so print out 5 4, 3, 2, 1
list2 = new_list.reverse() #It reverse again (to original), but returns none,
print(list2) #  so print None
new_list.reverse() # reverse from original 5, 4,3, 2, 1
list3 = new_list #give a reference to the above reversed sequence.
print(list3) # print out 5, 4, 3, 2, 1

new_list

new_list.sort()

None
[5, 4, 3, 2, 1]
None
[5, 4, 3, 2, 1]


### List methods

In [12]:
lst = [1,2,3]
lst.count(2) #Note just len(list) but also count the number of a specific element

1

Many times people find the difference between extend and append to be unclear. So note:**append: appends whole object at end:**

In [5]:
x = [1, 2, 3]
x.append([4, 5])
print(x)

[1, 2, 3, [4, 5]]


**extend: extends list by appending elements from the iterable:**

In [6]:
x = [1, 2, 3]
x.extend([4, 5])
print(x)

[1, 2, 3, 4, 5]


In [7]:
list1.index(2)

1

In [9]:
list1

[1, 2, 3, 4]

In [10]:
# Place a letter at the index 2
list1.insert(2,'inserted')

In [11]:
list1

[1, 2, 'inserted', 3, 4]

In [16]:
list1.remove('inserted') # removes the first occurrence of a value

In [17]:
list1

[1, 3, 4]

In [26]:
list2.sort(reverse=True)

### List comprehension
(1) in a square bracket for list.  
(2) expression + for loop + if statements 

In [ ]:
# Grab every letter in string
lst = [x for x in 'word']

# Square numbers in range and turn into list
lst = [x**2 for x in range(0,11)]

# Check for even numbers in a range
lst = [x for x in range(11) if x % 2 == 0]

# Convert Celsius to Fahrenheit
celsius = [0,10,20.1,34.5]

fahrenheit = [((9/5)*temp + 32) for temp in celsius ]

lst = [ x**2 for x in [x**2 for x in range(11)]]
lst

## Dictionary
### Dictionary Basics

In [2]:
d = {'key1':1,'key2':2,'key3':3}
d.keys()

dict_keys(['key1', 'key2', 'key3'])

In [3]:
d.values()

dict_values([1, 2, 3])

In [4]:
d.items()

dict_items([('key1', 1), ('key2', 2), ('key3', 3)])

In [1]:
{x:x**2 for x in range(10)}

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81}

One of the reasons it is not as common is the difficulty in structuring key names that are not based off the values.

**Dictionaries can be iterated over using the keys(), values() and items() methods.**

In [2]:
d = {'k1':1,'k2':2}

In [3]:
for k in d.keys():
    print(k)

k1
k2


In [4]:
for v in d.values():
    print(v)

1
2


In [5]:
for item in d.items():
    print(item)

('k1', 1)
('k2', 2)


By themselves the keys(), values() and items() methods return a dictionary *view object*. This is not a separate list of items. Instead, the view is always tied to the original dictionary.

In [6]:
key_view = d.keys()

key_view

dict_keys(['k1', 'k2'])

In [7]:
d['k3'] = 3

d

{'k1': 1, 'k2': 2, 'k3': 3}

In [8]:
key_view

dict_keys(['k1', 'k2', 'k3'])

### Dict comprehension

In [ ]:
new_fellowship = { member:len(member) for member in fellowship}

## set

**Creating non-empty set may be done with {} because set is a special dictionary. However, when creating an empty set, we need set() but not {}, as it will create an empty dictionary**.

In [1]:
s = set() 

In [2]:
s.add(1)

In [5]:
s.clear()

In [12]:
s = {1,2,3}
sc = s.copy()
sc.add(4)
print(s,sc)

{1, 2, 3} {1, 2, 3, 4}


In [22]:
print(s,sc)
s.difference(sc)
print(s.difference(sc)) # This is different from the next 
print(sc.difference(s))

{1, 3} {1, 2, 3, 4}
set()
{2, 4}


In [14]:
s1 = {1,2,3}

In [15]:
s2 = {1,4,5}

In [16]:
s1.difference_update(s2) #returns set1 after removing elements found in set2

In [17]:
s1

{2, 3}

In [17]:
s.discard(2)

In [18]:
s

{1, 3}

In [19]:
s1 = {1,2,3}

In [20]:
s2 = {1,2,4}

In [23]:
s1.intersection(s2)

{1, 2}

In [24]:
s1

{1, 2, 3}

In [25]:
s1.intersection_update(s2) #update a set with the intersection of itself and another.

In [26]:
s1

{1, 2}

In [27]:
s1 = {1,2}
s2 = {1,2,4}
s3 = {5}

s1.isdisjoint(s2) #return True if two sets have a null intersection

In [29]:
s1.isdisjoint(s3)

True

In [32]:
s1.issubset(s2)

True

In [33]:
s2.issuperset(s1)

True

In [34]:
s1.issuperset(s2)

False

In [35]:
s1

{1, 2}

In [36]:
s2

{1, 2, 4}

In [37]:
s1.symmetric_difference(s2) 
#Return the symmetric difference of two sets as a new set.
# i.e. all elements that are in exactly one of the sets.

{4}

In [38]:
s1.union(s2)

{1, 2, 4}

In [39]:
s1.update(s2) #Update a set with the union of itself and others

In [40]:
s1

{1, 2, 4}

## Tuple
### Tuple basics

In [6]:
t=('one', 2)

In [7]:
t.index('one')

0

In [8]:
t.count('one')

1

In [ ]:
## Set 
### Set basics

In [10]:
x = set() 
#set is like dic but with same key and values.
#So use the same {}

x.add(1)
x.add(2)
x.add(1)

In [11]:
x

{1, 2}

In [9]:
list1 = [1,1,2,2,3,4,5,6,1,1]

set(list1)

{1, 2, 3, 4, 5, 6}

## File

### Read-write file without appending

In [ ]:
%%writefile test.txt #This function is specific to jupyter notebooks.
Hello, this is a quick test file.

In [2]:
my_file = open('test.txt')

In [3]:
my_file.read()

'Hello, this is a quick test file.'

In [4]:
# But what happens if we try to read it again?
my_file.read()

''

In [5]:
# Seek to the start of file (index 0)
my_file.seek(0)

0

In [6]:
# Now read again
my_file.read()

'Hello, this is a quick test file.'

In [7]:
# Readlines returns a list of the lines in the file
my_file.seek(0)
my_file.readlines()

['Hello, this is a quick test file.']

In [8]:
my_file.close()

In [1]:
# Add a second argument to the function, 'w' which stands for write.
# Passing 'w+' lets us read and write to the file
my_file = open('test.txt','w+') 
#Opening a file with `'w'` or `'w+'` truncates the original, meaning that
#anything that was in the original file **is deleted**!

In [2]:
# Write to the file
my_file.write('This is a new line')

18

In [3]:
# Read the file
my_file.seek(0)
my_file.read()

'This is a new line'

In [4]:
my_file.close()  # always do this when you're done with a file

### Appending to a File
Passing the argument `'a'` opens the file and puts the pointer at the end, so anything written is appended. Like `'w+'`, `'a+'` lets us read and write to a file. If the file does not exist, one will be created.

In [5]:
my_file = open('test.txt','a+')
my_file.write('\nThis is text being appended to test.txt')
my_file.write('\nAnd another line here.')

23

In [6]:
my_file.seek(0)
print(my_file.read())

This is a new line
This is text being appended to test.txt
And another line here.


In [7]:
my_file.close()

We can do the same thing using IPython cell magic (%% is for cell magic functions): 

In [13]:
%%writefile -a test.txt

This is text being appended to test.txt
And another line here.

Appending to test.txt


### Iterating through without loading all contents to memory

Lets get a quick preview of a for loop by iterating over a text file. First let's make a new text file with some IPython Magic:

In [9]:
%%writefile test.txt
First Line
Second Line

Overwriting test.txt


In [10]:
for line in open('test.txt'): #line can be renamed to any strings
    print(line)

First Line

Second Line


# Built-in functions

## all() and any()

In [1]:
lst = [True,True,False,True]

In [2]:
all(lst)

False

In [3]:
any(lst)

True

## filter
The function filter(function,list) needs a function as its first argument. The function needs to return a **Boolean value**. This function will be applied to every element of the iterable. Only if the function returns True will the element of the iterable be included in the result. **Like map(), filter() returns an iterator**. 

However, when filter is defined in an object such as DataFrame, then it usually has only one parameter. See the Manipulating data frame with Pandas.  
by_com_filt = by_company.filter(lambda g:g['Units'].sum() > 35)  
**Check summary after map().**

In [1]:
def even_check(num):
    if num%2 ==0:
        return True

In [2]:
lst = range(20)
list(filter(even_check,lst))

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]

In [3]:
list(filter(lambda x: x%2 == 0,lst))

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]

## Lambda expression
syntax: lambda x, y... : f(x, y, ...)

In [19]:
lambda num: num ** 2 

<function __main__.<lambda>>

In [21]:
# You wouldn't usually assign a name to a lambda expression, this is just for demonstration!
square = lambda num: num **2

In [26]:
square(2)

4

** Lambda expression for grabbing the first character of a string: **


In [31]:
lambda s: s[0]

<function __main__.<lambda>>

** Lambda expression for reversing a string: **

In [32]:
lambda s: s[::-1]

<function __main__.<lambda>>

In [34]:
lambda x,y : x + y

<function __main__.<lambda>>

## map()
Unlike dictionary mapping keys to values, map() applies a function to iterable(s) in the form of map(function, iterable, ...) and returns an iterator. Using map with lambda expressions is much more common since the entire purpose of map() is to save effort on having to create manual for loops.  

**A summary**  
- filter() returns a boolean mask to filter another object. This is similar to other filtering masks used in e.g. DataFrame. Typical examples are df(df>0). However, apart from examples using 'df > 0' or dallas_1 = df['Destination Airport'] =='DAL' etc., we also use  dallas = df['Destination Airport'].str.contains('DAL') and other filtering with regular expression.  
- map() applies a function to an iterable. Function is only a special mapping and so the name map. Similar functions include .apply(), .transform(). 
- The above functions usually need two arguments when used as standalone function, and need one argument when defined on object. **filter and map actually are same in the original meaning of mapping.** 

In [8]:
def fahrenheit(celsius):
    return (9/5)*celsius + 32
temps = [0, 22.5, 40, 100]

F_temps = map(fahrenheit, temps) 
list(F_temps)

[32.0, 72.5, 104.0, 212.0]

In [ ]:
list(map(lambda x: (9/5)*x + 32, temps))

In [9]:
a = [1,2,3,4]
b = [5,6,7,8]
c = [9,10,11,12]

list(map(lambda x,y:x+y,a,b))

[6, 8, 10, 12]

In [10]:
list(map(lambda x,y,z:x+y+z,a,b,c))

[15, 18, 21, 24]

## range vs xrange vs arange

In Python 3, there is no xrange , but the range function in python 3 behaves like xrange in Python 2. If you want to write code that will run on both Python 2 and Python 3, you should use range(). That might be why that in python 3, range() returns a range object but not a list as in python 2.

xrange() – This function returns the generator object that can be used to display numbers only by looping. Only particular range is displayed on demand and hence called “lazy evaluation“.

arange belongs to numpy. It is mentioned here just for comparison.  

## range

In [1]:
range(0,11)

range(0, 11)

Note that this is a **generator** function, so to actually get a list out of it, we need to cast it to a list with **list()**. 

In [2]:
list(range(0,11,2))

[0, 2, 4, 6, 8, 10]

## reduce()
The function reduce(function, sequence) continually applies the function to the sequence. It then returns a single value. 

In [1]:
from functools import reduce

lst =[47,11,42,13]
reduce(lambda x,y: x+y,lst)

113

In [3]:
#Find the maximum of a sequence (This already exists as max())
max_find = lambda a,b: a if (a > b) else b

In [4]:
#Find max
reduce(max_find,lst)

47

## enumerate
Enumerate allows you to keep a count as you iterate through an object. It does this by returning a **tuple** in the form (count,element). 

In [13]:
index_count = 0

for letter in 'abc':
    print("At index {} the letter is {}".format(index_count,letter))
    index_count += 1

At index 0 the letter is a
At index 1 the letter is b
At index 2 the letter is c


Keeping track of how many loops you've gone through is so common, that enumerate was created so you don't need to worry about creating and updating this index_count or loop_count variable

In [14]:
# Notice the tuple unpacking!

for i,letter in enumerate('abc'):
    print("At index {} the letter is {}".format(i,letter))

At index 0 the letter is a
At index 1 the letter is b
At index 2 the letter is c


## zip

zip() returns an iterator of tuples that aggregates elements from each of the iterables.

In [13]:
mylist1 = [1,2,3,4,5]
mylist2 = ['a','b','c','d','e']

In [15]:
# This one is also a generator! We will explain this later, but for now let's transform it to a list
zip(mylist1,mylist2)

In [17]:
list(zip(mylist1,mylist2))

[(1, 'a'), (2, 'b'), (3, 'c'), (4, 'd'), (5, 'e')]

To use the generator, we could just use a for loop

In [20]:
for item1, item2 in zip(mylist1,mylist2):
    print('For this tuple, first item was {} and second item was {}'.format(item1,item2))

For this tuple, first item was 1 and second item was a
For this tuple, first item was 2 and second item was b
For this tuple, first item was 3 and second item was c
For this tuple, first item was 4 and second item was d
For this tuple, first item was 5 and second item was e


In [ ]:
#Finally lets use zip() to switch the keys and values of the two dictionaries
def switcharoo(d1,d2):
    dout = {}
    
    for d1key,d2val in zip(d1,d2.values()):
        dout[d1key] = d2val
    
    return dout

## in operator

We've already seen the **in** keyword durng the for loop, but we can also use it to quickly check if an object is in a list

In [21]:
'x' in ['x','y','z']

True

## random

Python comes with a built in random library. There are a lot of functions included in this random library, so we will only show you two useful functions for now.

In [18]:
from random import shuffle

In [19]:
# This shuffles the list "in-place" meaning it won't return
# anything, instead it will effect the list passed
mylist = [1,2,3,4,5]
shuffle(mylist)

In [20]:
mylist

[1, 4, 2, 3, 5]

In [39]:
from random import randint

In [41]:
# Return random integer in range [a, b], including both end points.
randint(0,100)

25

In [42]:
# Return random integer in range [a, b], including both end points.
randint(0,100)

91

## input

In [1]:
input('Enter Something into this box: ')

Enter Something into this box: Hello


'Hello'

# Miscellaneous important concepts
## Special function parameters

### **`*args` build a tuple**

In [2]:
def myfunc(*args):
    #print(type(args)) #The type of args is tuple, which is an iterable: __iter__ is defined but not __next__
    #print(args) # The print out results is (40,60,20). That is, a tuple. So the 40, 60, 20 passed into myfunc() will be 
    #transferred to a tuple, and then sum() accept the tuple. Tuple is just like a vector, or list, except it is immutable.
    return sum(args) * .05
    

print(myfunc(40,60,20))
#although args is a tuple, I cannot pass in a tuple like myfunc((40,60,20)). why? See comments above. sum() accepts tuple
#as arguments, but myfunc does not. It accept *+tuple. so the following is acceptable. 
myfunc(*(40,60,20))


6.0


6.0

### Understanding the asterisk(`*`) of Python-- unpacking the containers
https://medium.com/understand-the-python/understanding-the-asterisk-of-python-8b9daaa4a558  

The * can also be used for unpacking the containers. Its principles is similar to `*args` above. The easiest example is that we have data in the form of a list, tuple or dict, and a function take variable arguments:

In [4]:
from functools import reduce
primes = [2, 3, 5, 7, 11, 13]

def product(*numbers):
    print(type(numbers))
    print(numbers)

    p = reduce(lambda x, y: x * y, numbers)
    return p 

print(product(*primes))

print(product(primes))

# [2, 3, 5, 7, 11, 13]

<class 'tuple'>
(2, 3, 5, 7, 11, 13)
30030
<class 'tuple'>
([2, 3, 5, 7, 11, 13],)
[2, 3, 5, 7, 11, 13]


Because the product() take the variable arguments, we need to unpack the our list data and pass it to that function. In this case, if we pass the primes as *primes, every elements of the primes list will be unpacked, then stored in list called numbers. If pass that list primes to the function without unpacking, the numbers will has only one primes list not all elements of primes.
For tuple, it could be done exactly same to list, and for dict, just use ** instead of *.

### **`**kwargs` build a dictionary**

In [2]:
def myfunc(**kwargs):
    #print(type(kwargs)) #Note the type of kwargs are dict
    if 'fruit' in kwargs:
        print(f"My favorite fruit is {kwargs['fruit']}")  
        # review String Formatting and f-strings if this syntax is unfamiliar
    else:
        print("I don't like fruit")
        
myfunc(fruit='pineapple')

<class 'dict'>
My favorite fruit is pineapple


In [6]:
myfunc()

I don't like fruit


### **`*args` must be place before `**kwargs`**

In [7]:
def myfunc(*args, **kwargs):
    if 'fruit' and 'juice' in kwargs:
        print(f"I like {' and '.join(args)} and my favorite fruit is {kwargs['fruit']}")
        print(f"May I have some {kwargs['juice']} juice?")
    else:
        pass
        
myfunc('eggs','spam',fruit='cherries',juice='orange')

I like eggs and spam and my favorite fruit is cherries
May I have some orange juice?


## Variable scope, global and local


**LEGB Rule:**

L: Local — Names assigned in any way within a function (def or lambda), and not declared global in that function.

E: Enclosing function locals — Names in the local scope of any and all enclosing functions (def or lambda), from inner to outer.

G: Global (module) — Names assigned at the top-level of a module file, or declared global in a def within the file.

B: Built-in (Python) — Names preassigned in the built-in names module : open, range, SyntaxError,...

In [7]:
name = 'This is a global name'

def greet():
    # Enclosing function
    name = 'Sammy'
    
    def hello():
        print('Hello '+name)
    
    hello()

greet()

Hello hello


In [8]:
x = 50

def func(x):
    print('x is', x)
    x = 2
    print('Changed local x to', x)

func(x)
print('x is still', x)

x is 50
Changed local x to 2
x is still 50


In [9]:
x = 50

def func():
    global x #This is fine but not encouraged.
    print('This function is now using the global x!')
    print('Because of global x is: ', x)
    x = 2
    print('Ran func(), changed global x to', x)

print('Before calling func(), x is: ', x)
func()
print('Value of x (outside of func()) is: ', x)

Before calling func(), x is:  50
This function is now using the global x!
Because of global x is:  50
Ran func(), changed global x to 2
Value of x (outside of func()) is:  2


One last mention is that you can use the **globals()** and **locals()** functions to check what are your current local and global variables.


## With Statement Context Managers

When you open a file using `f = open('test.txt')`, the file stays open until you specifically call `f.close()`.  Should an exception be raised while working with the file, it remains open. This can lead to vulnerabilities in your code, and inefficient use of resources.

A context manager handles the opening and closing of resources, and provides a built-in `try/finally` block should any exceptions occur.

The best way to demonstrate this is with an example.

**Standard `open()` procedure, with a raised exception:**


In [1]:
p = open('oops.txt','a')
p.readlines()
p.close()

UnsupportedOperation: not readable

Let's see if we can modify our file:

In [2]:
p.write('add more text')

13

Ouch! I may not have wanted to do that until I traced the exception! Unfortunately, the exception prevented the last line, `p.close()` from running. Let's close the file manually:

In [3]:
p.close()

**Protect the file with `try/except/finally`**

A common workaround is to insert a `try/except/finally` clause to close the file whenever an exception is raised:


In [3]:
p = open('oops.txt','a')
try:
    p.readlines()
except:
    print('An exception was raised!')
finally:
    p.close()

An exception was raised!


Let's see if we can modify our file this time:

In [4]:
p.write('add more text')

ValueError: I/O operation on closed file.

Excellent! Our file is safe.
**Save steps with `with`**

Now we'll employ our context manager. The syntax follows `with [resource] as [target]: do something`

In [6]:
with open('oops.txt','a') as p:
    p.readlines()

UnsupportedOperation: not readable

Can we modify the file?

In [7]:
p.write('add more text')

ValueError: I/O operation on closed file.

Great! With just one line of code we've handled opening the file, enclosing our code in a `try/finally` block, and closing our file all at the same time.

## Errors and Exception Handling

In [1]:
try:
    f = open('testfile','w')
    f.write('Test write this')
except IOError:
    # This will only check for an IOError exception and then execute this print statement
    print("Error: Could not find file or read data")
else:
    print("Content written successfully")
    f.close()

Content written successfully


Now let's see what would happen if we did not have write permission (opening only with 'r'):

In [2]:
try:
    f = open('testfile','r')
    f.write('Test write this')
except IOError:
    # This will only check for an IOError exception and then execute this print statement
    print("Error: Could not find file or read data")
else:
    print("Content written successfully")
    f.close()

Error: Could not find file or read data


In [4]:
try:
    f = open('testfile','r')
    f.write('Test write this')
except: #we are not sure what exception is. 
    # This will check for any exception and then execute this print statement
    print("Error: Could not find file or read data")
else:
    print("Content written successfully")
    f.close()

Error: Could not find file or read data


In [5]:
try:
    f = open("testfile", "w")
    f.write("Test write statement")
    f.close()
finally: #If we want to run this code even when exception occured.
    print("Always execute finally code blocks")

Always execute finally code blocks


We can use this in conjunction with <code>except</code>. Let's see a new example that will take into account a user providing the wrong input:

In [6]:
def askint():
    try:
        val = int(input("Please enter an integer: "))
    except:
        print("Looks like you did not enter an integer!")

    finally:
        print("Finally, I executed!")
    print(val)

In [7]:
askint()

Please enter an integer: 5
Finally, I executed!
5


In [8]:
askint()

Please enter an integer: five
Looks like you did not enter an integer!
Finally, I executed!


UnboundLocalError: local variable 'val' referenced before assignment

Notice how we got an error when trying to print val (because it was never properly assigned). Let's remedy this by asking the user and checking to make sure the input type is an integer:

In [9]:
def askint():
    try:
        val = int(input("Please enter an integer: "))
    except:
        print("Looks like you did not enter an integer!")
        val = int(input("Try again-Please enter an integer: "))
    finally:
        print("Finally, I executed!")
    print(val)

In [10]:
askint()

Please enter an integer: five
Looks like you did not enter an integer!
Try again-Please enter an integer: four
Finally, I executed!


ValueError: invalid literal for int() with base 10: 'four'

Hmmm...that only did one check. How can we continually keep checking? We can use a while loop!

In [11]:
def askint():
    while True:
        try:
            val = int(input("Please enter an integer: "))
        except:
            print("Looks like you did not enter an integer!")
            continue
        else:
            print("Yep that's an integer!")
            break
        finally:
            print("Finally, I executed!")
        print(val)

In [12]:
askint()

Please enter an integer: five
Looks like you did not enter an integer!
Finally, I executed!
Please enter an integer: four
Looks like you did not enter an integer!
Finally, I executed!
Please enter an integer: 3
Yep that's an integer!
Finally, I executed!


So why did our function print "Finally, I executed!" after each trial, yet it never printed `val` itself? This is because with a try/except/finally clause, any <code>continue</code> or <code>break</code> statements **are reserved until after** the try clause is completed. This means that even though a successful input of **3** brought us to the <code>else:</code> block, and a <code>break</code> statement was thrown, the try clause continued through to <code>finally:</code> before breaking out of the while loop. And since <code>print(val)</code> was outside the try clause, the <code>break</code> statement prevented it from running.

Let's make one final adjustment:

In [13]:
def askint():
    while True:
        try:
            val = int(input("Please enter an integer: "))
        except:
            print("Looks like you did not enter an integer!")
            continue
        else:
            print("Yep that's an integer!")
            print(val)
            break
        finally:
            print("Finally, I executed!")

In [14]:
askint()

Please enter an integer: six
Looks like you did not enter an integer!
Finally, I executed!
Please enter an integer: 6
Yep that's an integer!
6
Finally, I executed!
